plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/editor/OpenDeclarationFinder.java

```diff
@@ -145,102 +145,98 @@ private void handleDescriptor(
-                                              .setSelectedRange(
-                                                  LinearRange.createWithStart(
-                                                          descriptor.getOffset())
-                                                      .andLength(0),
-                                                  true);
-                                          editor.activate();
-                                        }
-                                      }
-                                    });
-                          }
-                        });
-                  }
-                }
-              });
+      openFileFromWorkspace(descriptor);
     }
   }
+
+  private void activateOpenedEditor(
+      OpenDeclarationDescriptor descriptor, EditorPartPresenter openedEditor) {
+    editorAgent.openEditor(
+        openedEditor.getEditorInput().getFile(),
+        new OpenEditorCallbackImpl() {
+          @Override
+          public void onEditorOpened(EditorPartPresenter editor) {
+            setCursorAndActivateEditor(editor, descriptor.getOffset());
+          }
+
+          @Override
+          public void onEditorActivated(EditorPartPresenter editor) {
+            setCursorAndActivateEditor(editor, descriptor.getOffset());
+          }
+        });
+  }
+
+  private void openFileFromWorkspace(OpenDeclarationDescriptor descriptor) {
+    appContext
+        .getWorkspaceRoot()
+        .getFile(descriptor.getPath())
+        .then(
+            file -> {
+              if (file.isPresent()) {
+                openEditor(descriptor, file.get());
+              }
+            });
+  }
+
+  private void getLibraryEntry(Path projectPath, OpenDeclarationDescriptor descriptor) {
+    ExternalLibrariesParameters entryParams =
+        dtoFactory.createDto(ExternalLibrariesParameters.class);
+    entryParams.setNodeId(descriptor.getLibId());
+    entryParams.setNodePath(descriptor.getPath());
+    entryParams.setProjectUri(projectPath.toString());
+    extensionService
+        .libraryEntry(entryParams)
+        .then(
+            entry -> {
+              getBinaryContent(projectPath, descriptor, entry);
+            });
+  }
+
+  private void getBinaryContent(
+      Path projectPath, OpenDeclarationDescriptor descriptor, JarEntry entry) {
+    ExternalLibrariesParameters params = dtoFactory.createDto(ExternalLibrariesParameters.class);
+    params.setNodeId(descriptor.getLibId());
+    params.setNodePath(entry.getPath());
+    params.setProjectUri(projectPath.toString());
+    extensionService
+        .libraryNodeContentByPath(params)
+        .then(
+            content -> {
+              final VirtualFile file =
+                  javaNodeFactory.newJarFileNode(entry, descriptor.getLibId(), projectPath, null);
+              openEditor(descriptor, file);
+            });
+  }
+
+  private void openEditor(OpenDeclarationDescriptor descriptor, VirtualFile file) {
```

Comment:
Are you aware of the class "OpenFileInEditorHelper"? Maybe we can reuse stuff from there.

Reply:
@tsmaeder Do you really  think we should make some changes here? I mean in the feature this class will be removed because we have Find Definition feature. I've made here changes just for success build.

Reply:
true.